### PR TITLE
Use Anthology BibTeX

### PR DESCRIPTION
**Pull request by mjpost**
This PR intercepts the cache process to insert Anthology BibTeX for all papers that S2 crawled from the Anthology. It also refactors the code a bit, improving caching and timeouts.

Changes:
* BibTeX is now stored in the cache, since it requires a network query
* Allow multiply retries with exponential backoff
* The cache is always loaded and then written out completely after processing every author (helps save thousands of network queries)
* Entries are deleted from the cache if no longer present for an author on S2

Work-in-progress until the full crawl finishes.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1,39 +1,46 @@
...
 
-    if response.status_code != 200:
-        if first_attempt:
-            print("Error in request, got status code {0}".format(response.status_code))
-            time.sleep(60*5 + 30)
-            return return_request(request_type, request_id, first_attempt=False)
+        if response.status_code == 200:
+            break
+        elif response.status_code == 429:
+            sleep_time = 2 ** attempt_no
```
Comment (danyaljj) [inline main.py:52]: nice!

**Comment by danyaljj** [comment]
Looks great to me. Feel free to merge when the crawl finishes successfully.

**Comment by mjpost** [comment]
Thanks. Will update the cached results and new references. Note that the code now stores the bibtex in the cache, and loads and uses the cache every time.

**Comment by danyaljj** [comment]
One concern about caching: I am not sure how this code will play out with changes to S2 profiles. Two edge cases, in particular: 
 - Imagine if a paper gets picked up by the code and gets cached. Then I drop it from my S2 profile; will the code here also drop that paper? 
 - Another case: Imagine you publish an arxiv paper, and it gets fetched & cached. After a few months, you get it published. Would the code be able to ignore/revise the existing cache entry?

**Comment by mjpost** [comment]
Good thinking. I'll make sure it deletes from the cache. I'll have to see how S2 handles arxiv/published dupes.

**Comment by danyaljj** [comment]
Thanks! 

I suppose a naive way of handling is to never use cache. While the script would take much longer, I think it is not a bad idea given that the code will be run via a cronejob. So doesn't really matter whether it takes 2 minutes or 2 hours.

**Comment by mjpost** [comment]
Since the cache file is an argument you can always give it one that doesn't exist (say with the current date in it). But the cache is useful for testing at least. Still working on this, will ping you when done. Is there a deadline?

**Comment by danyaljj** [comment]
>  Still working on this, will ping you when done. Is there a deadline?

Sounds good. Not a hard deadline.

**Comment by danyaljj** [comment]
@mjpost let me know if there is anything I can help with so that we can get this to the finish line.

**Comment by mjpost** [comment]
lol/ouch. you said no hard deadline! :)

**Comment by mjpost** [comment]
(Give me through the end of the day...)

**Comment by mjpost** [comment]
Okay, I believe this is ready. I didn't address the dupes situation, but S2 does paper consolidation, so I think it is handled on their side.

The main thing is we now have Anthology BibTeX, so that's great.

**Comment by danyaljj** [comment]
Thank you so much, Matt!

**Pull request merged**